File: client/desktop/src/shared/types/bending-job.ts
// Backend BendingJob entity (Core/Entities/BendingJob.cs) yansıması.
// DataAPI POST /api/bending-jobs body'si bu şekli kabul eder.

import type {
  BendingDirection,
  BendingJobStatus,
  BendingMethod,
  ProfileType,
} from './enums'

export interface BendingJob {
  id: number

  // Lifecycle
  status: BendingJobStatus
  createdAt: string
  startedAt: string | null
  completedAt: string | null

  // Profile
  profileType: ProfileType
  direction: BendingDirection
  method: BendingMethod

  profileA: number
  profileB: number
  profileS: number
  targetDiameterMm: number
  profileH: number | null
  profileG: number | null

  // Part
  partLengthMm: number
  activeSensorSide: 'Left' | 'Right'
  valsCode: string | null

  // Bending parameters
  stepDistanceMm: number
  firstStepDistanceMm: number | null
  safetyMarginMm: number
  zeroResetDistanceMm: number
  pistonSpeedPercent: number
  rotationSpeedPercent: number
  slackDistanceMm: number | null
  slackPressureBar: number
  clampPressureBar: number
  toleranceMm: number

  // Machine geometry
  ballDiameterMm: number
  centerDistanceMm: number
  thetaDeg: number
  xA1: number
  yA1: number

  // Calculated
  calculatedPistonPositionMm: number | null

  // Progress
  totalPasos: number
  completedPasos: number
  currentPaso: number | null

  // Results
  measuredDiameterMm: number | null
  finalLeftPositionMm: number | null
  finalRightPositionMm: number | null
  durationSeconds: number | null

  // Error
  errorMessage: string | null
  failedAtPaso: number | null

  // User
  operatorName: string | null
  notes: string | null

  // Arc segments — server BendingJob.Segments nav property'sini include eder.
  // Arc dışı method'larda null/undefined. ArcNextSegmentModal kümülatif boy hesabı için okur.
  segments?: BendingSegmentInput[]
}

// Arc (çok açılı) bending — ilk segment payload'u (backend BendingSegment entity'si)
// Sonraki segment'ler POST /api/bending-job/{id}/next-segment ile interactive eklenir.
export interface BendingSegmentInput {
  segmentOrder: number  // 1 from UI (Create body); 2+ via interactive
  radiusMm: number      // yarıçap (UI'daki R direkt, /2 yok)
  angleDeg: number      // α — 0 < α < 180
  straightAfterMm: number  // L_düz, ≥ 0
}

// UI'dan job kaydederken gönderilecek minimum alanlar.
// Geri kalan field'lar backend default'larından gelir (DataAPI tarafında ileride netleştirilecek).
// Şu an için backend tüm alanları kabul ediyor; mapper bu interface'i full BendingJob'a çevirir.
export interface BendingJobCreateRequest {
  profileType: ProfileType
  direction: BendingDirection
  method: BendingMethod

  profileA: number
  profileB: number
  profileS: number
  targetDiameterMm: number
  profileH?: number | null
  profileG?: number | null

  partLengthMm: number
  activeSensorSide: 'Left' | 'Right'

  stepDistanceMm: number
  totalPasos?: number
  calculatedPistonPositionMm?: number | null

  // Arc-only fields (Method=Arc ise zorunlu; diğer method'larda null)
  totalSegmentCount?: number | null       // P
  arcStepDistanceMm?: number | null       // G — Arc'a özel job-level adım
  kivrimHizMetreDakika?: number | null    // H (m/min) — Arc'a özel hız
  segments?: BendingSegmentInput[]        // İlk segment burada; sonrakiler interactive
  // 2026-09-08: UI planner (ArcExtensionPlanner) ters sıralamayı önerdiyse true.
  // Segmentler DB'ye orijinal user-input sırada gider; runtime handler ters çevirir.
  isReversedArcOrder?: boolean | null

  // Sivama-only field (Method=Sivama ise zorunlu; diğer method'larda null)
  sivamaAngleDeg?: number | null          // X — Sıvama ilk rotasyon açısı (derece)

  // Operator metadata
  operatorName?: string | null
  notes?: string | null

  // İleride DataAPI default'lara bağlanırsa bu alanlar opsiyonel kalacak.
  safetyMarginMm?: number
  pistonSpeedPercent?: number
  rotationSpeedPercent?: number
  slackPressureBar?: number
  clampPressureBar?: number
  toleranceMm?: number
  ballDiameterMm?: number
  centerDistanceMm?: number
  thetaDeg?: number
  xA1?: number
  yA1?: number
  zeroResetDistanceMm?: number
}

// AI RECIPE FAZ 4B UI (2026-09-04) — bir dairesel bukum baslamadan onceki eslesme kontrolu
export interface SimilarBendingJobRequest {
  targetDiameterMm: number
  partLengthMm: number
  stepDistanceMm: number
  profileA: number
  profileB: number
  profileS: number
  ballDiameterMm: number
  // Opsiyonel — SignalR /machineHub'dan MachineState.Sensors.OilTempC verilirse ±2°C filtre uygulanir
  oilTempC?: number
}

export interface SimilarBendingJobIteration {
  iterationOrder: number
  measuredDiameterMm: number
  errorMm: number
  correctionPistonPositionMm: number
  measurementSide: string
  isWithinTolerance: boolean
}

export interface SimilarBendingJobResponse {
  // null = eslesme yok, ilk kez bu parametre kombinasyonu.
  match: {
    id: number
    durationSeconds: number
    totalSpringbackIterations: number
    createdAt: string
    completedAt: string
    oilTempAtStartC?: number | null
    oilTempAvgC?: number | null
    recipeUsedFromJobId?: number | null
  } | null
  iterations: SimilarBendingJobIteration[]
}

// Bending pipeline ilerlemesi (SignalR /machineHub "BendingProgress" event'i).
export interface BendingProgress {
  jobId: number
  completedPasos: number
  totalPasos: number
  percentComplete: number
  message: string

  // Arc interactive flow — FullCircle job'larında her zaman false/null.
  // awaitingArcSegmentInput=true iken UI SEKIL-14-LP modal'ı açıp R/α/L sorar,
  // POST /api/bending-jobs/{id}/segments ile DataApi'ye gönderir, modal kapanır.
  awaitingArcSegmentInput?: boolean
  completedSegmentOrder?: number | null
  totalSegmentCount?: number | null

  // Serpantin flow — pipeline yan dayama ayarı için operatör onayı bekliyor.
  // true iken UI büyük "DEVAM ET" butonu göster, tıklanınca
  // POST /api/bending-job/{id}/confirm-side-support çağrılır.
  awaitingSideSupportConfirmation?: boolean

  // Arc ölçüm hatası retry akışı (2026-08-18):
  //   awaitingMeasurementRetry=true iken UI MODAL 1 (Retract + Remeasure) açar.
  //   pneumaticRetractedForRetry=true olunca Modal 1'de "Tekrar Ölç" butonu AKTİF olur.
  //   awaitingMeasurementFinalDecision=true iken UI MODAL 2 (Finish + Skip) açar.
  //   POST /api/bending-job/{id}/measurement-retry-action ile action gönderilir.
  awaitingMeasurementRetry?: boolean
  pneumaticRetractedForRetry?: boolean
  awaitingMeasurementFinalDecision?: boolean
  prevMeasuredRadiusMm?: number | null
  lastMeasuredRadiusMm?: number | null
  targetRadiusMm?: number | null
  measurementRetrySegmentOrder?: number | null
  measurementRetryIteration?: number | null
}

File: engine-dataapi/CncBendingMachine.Core/Entities/BendingJob.cs
using CncBendingMachine.Core.Enums;

namespace CncBendingMachine.Core.Entities;

/// <summary>
/// Büküm iş kaydı - Electron UI'dan oluşturulur, büküm bu kayıt üzerinden takip edilir
/// </summary>
public class BendingJob
{
    public int Id { get; set; }

    // ============================================================
    // DURUM & YAŞAM DÖNGÜSÜ
    // ============================================================
    public BendingJobStatus Status { get; set; } = BendingJobStatus.Created;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? StartedAt { get; set; }
    public DateTime? CompletedAt { get; set; }

    // ============================================================
    // PROFİL PARAMETRELERİ (UI'dan - ŞEKİL-8, ŞEKİL-11)
    // ============================================================
    public ProfileType ProfileType { get; set; }
    public BendingDirection Direction { get; set; } = BendingDirection.Inward;
    public BendingMethod Method { get; set; } = BendingMethod.FullCircle;

    /// <summary>Profil yüksekliği A (mm) - 1. kenar ölçüsü</summary>
    public double ProfileA { get; set; }

    /// <summary>Profil genişliği B (mm) - 2. kenar ölçüsü</summary>
    public double ProfileB { get; set; }

    /// <summary>Et kalınlığı S (mm)</summary>
    public double ProfileS { get; set; }

    /// <summary>Hedef büküm çapı R (mm) - Ø değeri</summary>
    public double TargetDiameterMm { get; set; }

    /// <summary>Opsiyonel H parametresi (mm)</summary>
    public double? ProfileH { get; set; }

    /// <summary>Opsiyonel G parametresi (mm)</summary>
    public double? ProfileG { get; set; }

    // ============================================================
    // PARÇA PARAMETRELERİ
    // ============================================================

    /// <summary>Parça uzunluğu (mm)</summary>
    public double PartLengthMm { get; set; }

    /// <summary>Aktif sensör tarafı: "Left" veya "Right"</summary>
    public string ActiveSensorSide { get; set; } = "Left";

    /// <summary>Vals topu kodu (ŞEKİL-11 alt kısım)</summary>
    public string? ValsCode { get; set; }

    // ============================================================
    // BÜKÜM PARAMETRELERİ
    // ============================================================

    /// <summary>Paso adım mesafesi (mm)</summary>
    public double StepDistanceMm { get; set; } = 30;

    /// <summary>İlk paso adım mesafesi (mm) - opsiyonel</summary>
    public double? FirstStepDistanceMm { get; set; }

    /// <summary>Güvenlik payı (mm)</summary>
    public double SafetyMarginMm { get; set; } = 50;

    /// <summary>Parça sıfırlama mesafesi (mm) — sensörden "0" noktasına (DB ayarlarından snapshot)</summary>
    public double ZeroResetDistanceMm { get; set; } = 690;

    /// <summary>Piston hızı (%)</summary>
    public int PistonSpeedPercent { get; set; } = 100;

    /// <summary>Rotasyon hızı (%)</summary>
    public int RotationSpeedPercent { get; set; } = 100;

    /// <summary>Boşluk alma mesafesi (mm) - opsiyonel</summary>
    public double? SlackDistanceMm { get; set; }

    /// <summary>Boşluk alma basıncı (bar)</summary>
    public int SlackPressureBar { get; set; } = 157;

    /// <summary>Parça sıkıştırma basıncı (bar) — Clamp adımı ve auto-correct içinde kullanılır</summary>
    public int ClampPressureBar { get; set; } = 155;

    /// <summary>Tolerans (mm) - ŞEKİL-14 sağ üst</summary>
    public double ToleranceMm { get; set; } = 0.1;

    // ============================================================
    // MAKİNE PARAMETRELERİ (hesaplama için)
    // ============================================================

    /// <summary>Vals topu çapı (mm)</summary>
    public double BallDiameterMm { get; set; } = 220;

    /// <summary>Top eksen mesafesi (mm)</summary>
    public double CenterDistanceMm { get; set; } = 300.82;

    /// <summary>Sabit açı (derece)</summary>
    public double ThetaDeg { get; set; } = 63;

    /// <summary>Sol top X koordinatı (mm)</summary>
    public double XA1 { get; set; } = -493;

    /// <summary>Sol top Y koordinatı (mm)</summary>
    public double YA1 { get; set; } = 0;

    // ============================================================
    // HESAPLANAN DEĞERLER (start sırasında doldurulur)
    // ============================================================

    /// <summary>Hesaplanan piston hedef pozisyonu (mm)</summary>
    public double? CalculatedPistonPositionMm { get; set; }

    // ============================================================
    // İLERLEME TAKİBİ
    // ============================================================

    /// <summary>Toplam paso sayısı</summary>
    public int TotalPasos { get; set; }

    /// <summary>Tamamlanan paso sayısı</summary>
    public int CompletedPasos { get; set; }

    /// <summary>Şu anki paso (çalışırken)</summary>
    public int? CurrentPaso { get; set; }

    // ============================================================
    // SONUÇLAR
    // ============================================================

    /// <summary>Ölçülen çap (mm) - geri esneme sensöründen</summary>
    public double? MeasuredDiameterMm { get; set; }

    /// <summary>Son sol piston pozisyonu (mm)</summary>
    public double? FinalLeftPositionMm { get; set; }

    /// <summary>Son sağ piston pozisyonu (mm)</summary>
    public double? FinalRightPositionMm { get; set; }

    /// <summary>Toplam süre (saniye)</summary>
    public double? DurationSeconds { get; set; }

    // ============================================================
    // HATA BİLGİSİ
    // ============================================================

    /// <summary>Hata mesajı (fail durumunda)</summary>
    public string? ErrorMessage { get; set; }

    /// <summary>Hangi pasoda fail oldu</summary>
    public int? FailedAtPaso { get; set; }

    // ============================================================
    // KULLANICI BİLGİSİ
    // ============================================================

    /// <summary>Operatör adı</summary>
    public string? OperatorName { get; set; }

    /// <summary>Notlar</summary>
    public string? Notes { get; set; }
}
